### PR TITLE
Install DALI within conda environment

### PR DESCRIPTION
**Pull request by szalpal**
DALI Backend links to DALI using DALI wheel. Currently, the wheel is downloaded in the CMake (via `pip download`) and unzipped into proper location. For the autoserialize feature (#125), simple downloading is no longer sufficient, as we need to run DALI's Python API from the C++*. The solution is to install the wheel in the target docker image. However, if the DALI wheel would be brutally installed by the root in the `tritonserver` image, there might occur a situation, when dependencies between multiple wheels are incompatible with each other.

This PR proposes, how to install the DALI wheel as the conda environment. The build variables remain the same, functionality is maintained. The PR only adds the wheel download+install. Proper activating conda environment (for the purpose of triggering Python script from C++) will be included in next PR.

**Important:**
- [x] Make sure, that the `conda` (most probably in miniconda flavour) is available in the `server` build environment (i.e. a builder docker image). https://github.com/triton-inference-server/server/pull/4476

\* That's because the autoserialize is based on the Python code. And DALI's C/C++ API aren't able to serialize the pipeline.

### CODE REVIEW

**Comment by szalpal** [comment]
Note to reviewers: Yes, I'm aware of missing newlines at the end of files. I'll fix it along with applying review remarks.

File: docker/Dockerfile.devel
Comment (JanuszL) [inline docker/Dockerfile.devel:72]: Do you need this shasum?

File: docker/Dockerfile.release
Comment (JanuszL) [inline docker/Dockerfile.release:81]: What is the purpose of this set of commands?

File: src/dali_executor/CMakeLists.txt
Comment (JanuszL) [inline src/dali_executor/CMakeLists.txt:42]: I wonder if we need this additional level of indirection through dalienv.yml, how about:
```suggestion
        conda create -n dalienv python=3.8 -y && conda install -n dalienv -y pip && conda run -n dalienv pip install --extra-index-url ${DALI_EXTRA_INDEX_URL} ${EXTRA_DOWNLOAD_ARGS} ${DALI_DOWNLOAD_PKG_NAME}
```

File: src/dali_executor/CMakeLists.txt
Comment (JanuszL) [inline src/dali_executor/CMakeLists.txt:50]: Maybe we should not hardcode it but something like:
```
conda run -n dalienv python -c "import nvidia.dali.sysconfig; print(nvidia.dali.sysconfig.get_include_dir()"
```
and
```
conda run -n dalienv python -c "import nvidia.dali.sysconfig; print(nvidia.dali.sysconfig.get_lib_dir()"
```

File: docker/Dockerfile.devel
Comment (JanuszL) [inline docker/Dockerfile.devel:77]: How it works with the TRION releases process? Does the TRITON image have conda env?

File: docker/Dockerfile.release
Comment (prak-nv) [inline docker/Dockerfile.release:73]: Wouldn't you want something like ``sha256sum -c ./shasum && \`` here to really check sha before running it?

File: docker/Dockerfile.devel
Comment (szalpal) [reply docker/Dockerfile.devel:72]: FTR: This whole snippet which installs the Miniconda, I took from `continuumio/miniconda3` docker image:
https://github.com/ContinuumIO/docker-images/blob/master/miniconda3/debian/Dockerfile

I removed some of the unnecessary commands.

And answering your question: I don't suppose I need this, but I figured that it might be nice to check the install script checksum. If you feel that it's unnecessary, I'll remove it.

File: docker/Dockerfile.release
Comment (szalpal) [reply docker/Dockerfile.release:81]: Good question. The first two I believe remove some build/install artifacts. The last one, from the conda help:

```
$ conda clean --help

Remove unused packages and caches.

  -a, --all             Remove index cache, lock files, unused cache packages, and tarballs.
  -f, --force-pkgs-dirs
                        Remove *all* writable package caches. This option is not included with the --all flag. WARNING: This will break environments with packages installed using symlinks back to the package cache.
  -y, --yes             Do not ask for confirmation.
```

File: src/dali_executor/CMakeLists.txt
Comment (szalpal) [reply src/dali_executor/CMakeLists.txt:42]: I believe both would work equally well, both have pros and cons. @prak-nv , what your opinion on this?

File: src/dali_executor/CMakeLists.txt
Comment (szalpal) [reply src/dali_executor/CMakeLists.txt:50]: Good point. I'll change this.

File: docker/Dockerfile.devel
Comment (szalpal) [reply docker/Dockerfile.devel:77]: I believe, that Triton does not have conda in the build environment. I've put this requirement in the description of the PR. I won't merge this, unless Triton has conda in the build environment.

File: docker/Dockerfile.release
Comment (szalpal) [reply docker/Dockerfile.release:73]: Good point, added.

File: src/dali_executor/CMakeLists.txt
Comment (prak-nv) [reply src/dali_executor/CMakeLists.txt:42]: @szalpal version seems to be more readable/explicit for me.

Comment (JanuszL) [reply src/dali_executor/CMakeLists.txt:42]: I'm fine with both.

File: docker/Dockerfile.devel
Comment (JanuszL) [reply docker/Dockerfile.devel:72]: Up to you.

Comment (prak-nv) [inline docker/Dockerfile.devel:72]: You'll also need ssh256sum here. Sorry, didn't notice this in previous round

File: src/dali_executor/CMakeLists.txt
Comment (JanuszL) [reply src/dali_executor/CMakeLists.txt:50]: So maybe you should use `execute_process` instead of `add_custom_command` ?
Is there any value in making the creation of the conda env a build step instead of part of the configuration process?

Comment (szalpal) [reply src/dali_executor/CMakeLists.txt:50]: The value is that it's conducted once per target build. In a hypothetical scenario, that somebody goes:
```
cmake -D TRITON_DALI_SKIP_DOWNLOAD=OFF ..
make
```
And then, during the development touches any of the `CMakeLists.txt` files, consecutive `make` command won't re-download and re-create the conda env. Which is forbidden - user would need to remove entire `build` directory and run `cmake` from scratch.

Comment (szalpal) [reply src/dali_executor/CMakeLists.txt]: Tried this one. This also has hard time working, because of `configure_file` command above: we need the configured file to create conda environment. OTOH, `configure_file` happens at build stage, with all its repercussions, e.g. if the target is not linked to any executable/library, the file won't be generated. This makes the commands that rely on `dalienv.yml` no-op in the generation step (because this file doesn't exist yet).

Judging all the tweaks here, I believe it is better to hardcode the library paths. It in an easy error to spot - the build will break if there are no libraries at that location, and an easy error to fix if necessary - if the library paths will change, we just need to hardcode different path.

Also, this is not really a technical point, but my gut (and slight experience with CMake) tells me that the build stage is better suitable for workloads such as downloading dependencies. An argument for this might be, that it happens concurrently - when we go `make -j`, one thread handles the download, while everything else that does not depend on the downloading target might be built simultaneously.

Taking into account all this, I believe it would be better to keep creating the conda environment in the build stage, which (unfortunately) stipulates hardcoding the library paths. But, as said earlier, resolving error that might occur here would be rather easy.

Comment (JanuszL) [reply src/dali_executor/CMakeLists.txt:50]: Up to you.

> Also, this is not really a technical point, but my gut (and slight experience with CMake) tells me that the build stage is better suitable for workloads such as downloading dependencies. An argument for this might be, that it happens concurrently - when we go make -j, one thread handles the download, while everything else that does not depend on the downloading target might be built simultaneously.

To make sure that it works as intended you need to add a dependency between the DALI backed binaries and `dali_env`. Otherwise, it will attempt to use includes from the env that is just created in parallel.

Comment (JanuszL) [reply src/dali_executor/CMakeLists.txt:50]: ..., ok, I see L63. NVM then.

**Review by JanuszL — APPROVED** [review]

**Review by prak-nv — APPROVED** [review]

**Pull request merged**
